R/spark_shell.R

```diff
@@ -120,79 +134,110 @@ start_shell <- function(master,
   # add sparkr-shell to args
   shell_args <- c(shell_args, app_jar)
 
-  # create temporary file for shell ports output and add it to the args
-  shell_output_path <- spark_config_value(config,
-                                          "sparklyr.ports.file",
-                                          normalizePath(tempfile(fileext = ".out"),
-                                                        mustWork = FALSE))
+  gatewayPort <- as.integer(spark_config_value(config, "sparklyr.gateway.port", "8880"))
+  isService <- as.logical(spark_config_value(config, "sparklyr.service", "FALSE"))
+  sessionId <- floor(runif(1, min=0, max=10000))
+  shell_args <- c(shell_args, as.character(gatewayPort), sessionId)
+  gateway <- NULL
+  if (isService) {
+    shell_args <- c(shell_args, "--service")
 
-  on.exit(unlink(shell_output_path))
-  shell_args <- c(shell_args, shell_output_path)
+    # try connecting to an existing instance of sparklyr running in service mode
+    gateway <- wait_connect_gateway(gatewayPort, 1)
+  }
 
-  # create temp file for stdout and stderr
-  output_file <- tempfile(fileext = "_spark.log")
-  error_file <- tempfile(fileext = "_spark.err")
+  output_file <- NULL
+  error_file <- NULL
+
+  if (is.null(gateway))
+  {
+    # create temp file for stdout and stderr
+    output_file <- tempfile(fileext = "_spark.log")
+    error_file <- tempfile(fileext = "_spark.err")
+
+    # start the shell (w/ specified additional environment variables)
+    env <- unlist(environment)
+    withr::with_envvar(env, {
+      if (.Platform$OS.type == "windows") {
+        shell(paste(
+          spark_submit_path,
+          paste(shell_args, collapse = " "),
+          ">",
+          output_file,
+          "2>",
+          error_file
+        ),
+        wait = FALSE)
+      }
+      else {
+        system2(spark_submit_path,
+          args = shell_args,
+          stdout = output_file,
+          stderr = output_file,
+          wait = FALSE)
+      }
+    })
 
-  # start the shell (w/ specified additional environment variables)
-  env <- unlist(environment)
-  withr::with_envvar(env, {
-    if (.Platform$OS.type == "windows") {
-      shell(paste(
+    # wait for the service to start
+    waitSeconds <- spark_config_value(config, "sparklyr.gateway.wait.seconds", 10)
+    gateway <- wait_connect_gateway(gatewayPort, waitSeconds)
+
+    if (is.null(gateway)) {
+      abort_shell(
+        "Failed while connecting to sparklyr",
         spark_submit_path,
-        paste(shell_args, collapse = " "),
-        ">",
+        shell_args,
         output_file,
-        "2>",
         error_file
-      ),
-      wait = FALSE)
+      )
     }
-    else {
-      system2(spark_submit_path,
-              args = shell_args,
-              stdout = output_file,
-              stderr = output_file,
-              wait = FALSE)
+  }
+
+  tryCatch({
+    readInt <- function(con, n = 1) {
```

Comment:
We already have `readInt` defined in `sparklyr`; do we need to redefine it here?